### PR TITLE
style: Move all origin-specific cascade data to PerOriginCascadeData

### PR DESCRIPTION
**Pull request by heycam**
This is a preliminary refactoring in preparation for only rebuilding cascade data for origins that have a change. https://bugzilla.mozilla.org/show_bug.cgi?id=1382925

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/18017)

### CODE REVIEW

File: components/style/stylist.rs [hunk 1]
```diff
@@ -1635,6 +1635,8 @@ impl CascadeData {
     }
 }
 
+/// Iterator over `PerOriginCascadeData`, from highest level (user) to lowest
+/// (user agent).
```
Comment (emilio) [inline components/style/stylist.rs]: Add a comment to note that we rely on that to get animations, and potentially other stuff?

File: components/selectors/bloom.rs [hunk 2]
```diff
@@ -179,9 +136,84 @@ impl BloomFilter {
...
+        hash1(hash) as usize
+    }
+
+    #[inline]
+    fn second_slot_index(hash: u32) -> usize {
+        hash2(hash) as usize
+    }
+}
+
+pub struct BloomStorageU8 {
```
Comment (emilio) [inline components/selectors/bloom.rs]: needs docs.

File: components/selectors/bloom.rs [hunk 3]
```diff
@@ -216,6 +221,55 @@ impl Clone for BloomStorageU8 {
...
+                "should not decrement if slot is already false");
+
+        if increment {
+            *byte |= bit;
+        }
+    }
+
+    #[inline]
+    fn slot_is_empty(&self, index: usize) -> bool {
+        let bit = 1 << (index % 8);
```
Comment (emilio) [inline components/selectors/bloom.rs]: nit: maybe create a method for this? presumably not worth the effort.

File: components/selectors/bloom.rs [hunk 4]
```diff
@@ -179,9 +136,84 @@ impl BloomFilter {
     }
 }
 
-#[inline]
-fn full(slot: &u8) -> bool {
-    *slot == 0xff
+pub trait BloomStorage : Clone + Default {
+    fn slot_is_empty(&self, index: usize) -> bool;
+    fn adjust_slot(&mut self, index: usize, increment: bool);
```
Comment (emilio) [inline components/selectors/bloom.rs]: Could you mention what are these functions expected to do?

I found `adjust` to be a little not too descriptive... But I can't think of a better name that isn't super-long, so, maybe something simple like:

```
Either increment or decrement the slot at `index`.
```

Would be helpful.

File: components/selectors/bloom.rs [hunk 5]
```diff
@@ -237,6 +291,10 @@ fn hash2(hash: u32) -> u32 {
 fn create_and_insert_some_stuff() {
     let mut bf = BloomFilter::new();
 
+    assert!(ARRAY_SIZE % 8 == 0,
```
Comment (emilio) [inline components/selectors/bloom.rs]: I think you can static-assert this with transmute or something like that...

```
transmute::<[u8; 1 + ARRAY_SIZE % 8], [u8; 1]>(..)
```

File: components/style/stylist.rs [hunk 6]
```diff
@@ -1683,7 +1686,9 @@ impl PerOriginCascadeData {
     }
 
     fn clear(&mut self) {
-        *self = Self::new();
+        self.element_map = SelectorMap::new();
```
Comment (emilio) [inline components/style/stylist.rs]: Is this change needed?

File: components/style/stylist.rs [hunk 7]
```diff
@@ -539,13 +534,13 @@ impl Stylist {
 
                     // Don't let a prefixed keyframes animation override a non-prefixed one.
                     let needs_insertion = keyframes_rule.vendor_prefix.is_none() ||
-                        self.animations.get(keyframes_rule.name.as_atom()).map_or(true, |rule|
+                        origin_cascade_data.animations.get(keyframes_rule.name.as_atom()).map_or(true, |rule|
```
Comment (emilio) [inline components/style/stylist.rs]: While you're here, the indentation here is a bit weird, maybe break it up a bit, or add braces around the closure body? (If you feel like it)

File: components/style/stylist.rs [hunk 8]
```diff
@@ -269,9 +265,23 @@ impl Stylist {
         self.selectors_for_cache_revalidation.len()
     }
 
-    /// Gets a reference to the invalidation map.
-    pub fn invalidation_map(&self) -> &InvalidationMap {
-        &self.invalidation_map
+    /// Returns the number of entries in invalidation maps.
+    pub fn num_invalidations(&self) -> usize {
+        self.cascade_data.iter_origins()
+            .map(|d| d.invalidation_map.len()).sum()
```
Comment (emilio) [inline components/style/stylist.rs]: nit: I'd move `sum`, and maybe `iter_origins`, to its own line, but no huge deal.

File: components/style/stylist.rs [hunk 9]
```diff
@@ -578,9 +543,18 @@ impl Stylist {
             // we rebuild.
             true
         } else if *local_name == local_name!("style") {
-            self.style_attribute_dependency
+            self.cascade_data
+                .iter_origins()
+                .map(|d| d.style_attribute_dependency)
```
Comment (emilio) [inline components/style/stylist.rs]: nit: just `any(|d| d.style_attribute_dependency)`.

File: components/style/stylist.rs [hunk 10]
```diff
@@ -578,9 +543,18 @@ impl Stylist {
...
+                .any(|b| b)
         } else {
-            self.attribute_dependencies.might_contain_hash(local_name.get_hash())
+            self.cascade_data
+                .iter_origins()
+                .map(|d| {
+                    d.attribute_dependencies
+                        .might_contain_hash(local_name.get_hash())
+                })
+                .any(|b| b)
```
Comment (emilio) [inline components/style/stylist.rs]: ditto.

File: components/style/stylist.rs [hunk 11]
```diff
@@ -592,14 +566,17 @@ impl Stylist {
...
     }
 
     /// Returns whether the given ElementState bit is relied upon by a selector
     /// of some rule in the stylist.
     pub fn has_state_dependency(&self, state: ElementState) -> bool {
-        self.state_dependencies.intersects(state)
+        self.cascade_data
+            .iter_origins()
+            .map(|d| d.state_dependencies.intersects(state))
+            .any(|b| b)
```
Comment (emilio) [inline components/style/stylist.rs]: ditto

File: components/style/stylist.rs [hunk 12]
```diff
@@ -1319,7 +1296,10 @@ impl Stylist {
     /// of our rule maps.
     #[inline]
     pub fn may_have_rules_for_id(&self, id: &Atom) -> bool {
-        self.mapped_ids.might_contain_hash(id.get_hash())
+        self.cascade_data
+            .iter_origins()
+            .map(|d| d.mapped_ids.might_contain_hash(id.get_hash()))
+            .any(|b| b)
```
Comment (emilio) [inline components/style/stylist.rs]: ditto

**Review by emilio — APPROVED** [review]
Looks pretty straight-forward. r=me with the nits addressed, thanks!

File: components/style/stylist.rs (hunk 6 above)
Comment (heycam) [inline components/style/stylist.rs]: Yes, later additions to PerOriginCascadeData will want to use `clear()` on some members (e.g. the Bloom filter ones).

File: components/selectors/bloom.rs (hunk 3 above)
Comment (heycam) [inline components/selectors/bloom.rs]: Yeah, I almost did, then decided it wasn't worth it.

File: components/selectors/bloom.rs [hunk 13]
```diff
@@ -179,9 +136,85 @@ impl BloomFilter {
...
+/// Storage class for a CountingBloomFilter that has 8-bit counters.
+pub struct BloomStorageU8 {
+    counters: [u8; ARRAY_SIZE],
+}
+
+impl BloomStorage for BloomStorageU8 {
+    #[inline]
+    fn adjust_slot(&mut self, index: usize, increment: bool) {
+        let slot = &mut self.counters[index];
+        if *slot == 0xff {  // full
```
Comment (emilio) [inline components/selectors/bloom.rs]: Err, this needs to be `!=`. Travis is red because of this.

Comment (heycam) [inline components/selectors/bloom.rs]: That is probably why my try run is all orange. :-)
